controllers/openstackcluster_controller_test.go

```diff
@@ -0,0 +1,224 @@
+				Kind:       "Cluster",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      capiClusterName,
+				Namespace: testNamespace,
+			},
+		}
+
+		input := framework.CreateNamespaceInput{
+			Creator: k8sClient,
+			Name:    testNamespace,
+		}
+		framework.CreateNamespace(ctx, input)
+	})
+
+	AfterEach(func() {
+		orphan := metav1.DeletePropagationOrphan
+		deleteOptions := client.DeleteOptions{
+			PropagationPolicy: &orphan,
+		}
+
+		// Remove finalizers and delete openstackcluster
+		patchHelper, err := patch.NewHelper(testCluster, k8sClient)
+		Expect(err).To(BeNil())
+		testCluster.SetFinalizers([]string{})
+		err = patchHelper.Patch(ctx, testCluster)
+		Expect(err).To(BeNil())
+		err = k8sClient.Delete(ctx, testCluster, &deleteOptions)
+		Expect(err).To(BeNil())
+		// Remove finalizers and delete cluster
+		patchHelper, err = patch.NewHelper(capiCluster, k8sClient)
+		Expect(err).To(BeNil())
+		capiCluster.SetFinalizers([]string{})
+		err = patchHelper.Patch(ctx, capiCluster)
+		Expect(err).To(BeNil())
+		err = k8sClient.Delete(ctx, capiCluster, &deleteOptions)
+		Expect(err).To(BeNil())
+		input := framework.DeleteNamespaceInput{
+			Deleter: k8sClient,
+			Name:    testNamespace,
+		}
+		framework.DeleteNamespace(ctx, input)
+	})
+
+	It("should do nothing when owner is missing", func() {
+		testCluster.SetName("missing-owner")
+		testCluster.SetOwnerReferences([]metav1.OwnerReference{})
+
+		err := k8sClient.Create(ctx, testCluster)
+		Expect(err).To(BeNil())
+		err = k8sClient.Create(ctx, capiCluster)
+		Expect(err).To(BeNil())
+		req := createRequestFromOSCluster(testCluster)
+
+		result, err := reconciler.Reconcile(ctx, req)
+		// Expect no error and empty result
+		Expect(err).To(BeNil())
+		Expect(result).To(Equal(reconcile.Result{}))
+	})
+	It("should do nothing when paused", func() {
+		testCluster.SetName("paused")
+		annotations.AddAnnotations(testCluster, map[string]string{clusterv1.PausedAnnotation: "true"})
+
+		err := k8sClient.Create(ctx, testCluster)
+		Expect(err).To(BeNil())
+		err = k8sClient.Create(ctx, capiCluster)
+		Expect(err).To(BeNil())
+		req := createRequestFromOSCluster(testCluster)
+
+		result, err := reconciler.Reconcile(ctx, req)
+		// Expect no error and empty result
+		Expect(err).To(BeNil())
+		Expect(result).To(Equal(reconcile.Result{}))
+	})
+	It("should do nothing when unable to get OS client", func() {
+		testCluster.SetName("no-openstack-client")
+		err := k8sClient.Create(ctx, testCluster)
+		Expect(err).To(BeNil())
+		err = k8sClient.Create(ctx, capiCluster)
+		Expect(err).To(BeNil())
+		req := createRequestFromOSCluster(testCluster)
+
+		result, err := reconciler.Reconcile(ctx, req)
+		// Expect error for getting OS clinet and empty result
+		Expect(err).ToNot(BeNil())
+		Expect(result).To(Equal(reconcile.Result{}))
+	})
+
+	// TODO: This test is set to pending (PIt instead of It) since it is not working.
+	PIt("should be able to reconcile when basition disabled", func() {
```

Comment:
Please take a look at this test specifically. I'm not sure how to move forward here. The basic tests above does not require any interaction with the Openstack API, but this one does and I didn't find a good way to fake it. Am I missing something? Is it possible to get a fake/mock scope or dummy openstack API to use in tests like these?